README.md
backup-s3
=========

__backup-s3__ periodically backs up a database (postgres or mysql) and/or data folder(s) to Amazon S3. Backups can be restored with a single command.

New in v3.0: it is now possible to target a specific version of postgres client from v13 till v15
New in v2.2: it is now possible to specify the S3 endpoint to work with any alternative S3-compatible endpoint.

The typical use case is to back up an application (e.g. Wordpress) run with docker-compose.

![backup-s3.png](https://bitbucket.org/hleroy/backup-s3/raw/67be5a9d7b52717f70dbe7653c8f51ce00ce789d/backup-s3.png)


Supported tags
--------------

* v3.0-pg15, latest
* v3.0-pg14
* v3.0-pg13


Setup with docker-compose
-------------------------

Add backup-s3 to your compose file. This is an example to backup a Wordpress container:

```yaml
  backup:
    image: hleroy/backup-s3
    init: true                   # Enable faster container termination
    volumes:
      - wordpress:/var/www/html  # Volume to back up
    environment:                 # Provide env secrets (S3 key etc...)
      BACKUP_ENABLED: yes        # Any non-empty value will work
      # S3
      S3_REGION: eu-central-1
      S3_BUCKET: my-bucket
      S3_ACCESS_KEY_ID: access_key_id
      S3_SECRET_ACCESS_KEY: secret_access_key
      # S3_ENDPOINT_URL:         # Optional: only if you are using an alternative S3-compatible endpoint
      # Database
      DB_ENGINE: mysql
      DB_NAME: wordpress
      DB_USER: wordpress
      DB_PASS: wordpress
      DB_HOST: mysql
      DB_PORT: 5432
      # Data
      DATA_PATH: '/var/www/html' # Customize to your needs
      # Cron schedule
      CRON_SCHEDULE: '0 0 * * *' # Every day at midnight
```

`BACKUP_ENABLED` (new in v2.0) is mandatory. Any non-empty value will work (e.g. True, Yes). If omitted, the container exits with code 0.

S3 settings (`S3_REGION`, `S3_BUCKET`, `S3_ACCESS_KEY_ID`, `S3_SECRET_ACCESS_KEY`) are all mandatory. If any of them is missing, the container exits with code 1.

`S3_ENDPOINT_URL` (new in v2.2) is optional in case you wish to use an alternative S3-compatible endpoint. e.g. for Scaleway Paris region `https://s3.fr-par.scw.cloud`

`DB_ENGINE` can be `mysql` or `postgres`. If omitted, no database backup is attempted. If present, but database settings are missing, the container exits with code 1.
`DB_PORT` is optional. It defaults to 3306 (for mysql) or 5432 (for postgres).

`DATA_PATH` must be an absolute path with no trailing slash. If omitted, no data backup is attempted. It is possible to specify multiple paths separated by a colon (:), e.g. ```DATA_PATH: '/var/www/html/images:/var/www/html/extensions:/var/www/data'```

`CRON_SHEDULE` uses the usual cron format. Try https://crontab.guru/ to customize to your needs. If omitted, it defaults to `0 0 * * *` (every day at midnight).


Usage
-----

When your Compose application is started, backup will be run according to the CRON schedule.

Backup filenames have a timestamp and an appended random string.
E.g.  `2020-04-22T00:00:01Z.87793a5e1556e0ff8a8b528328fab4ad.tgz`

You can run an immediate backup by running:
`docker-compose run --rm backup no-cron`

You can restore a specific backup by running:
`docker-compose run --rm backup restore BACKUP_FILENAME`

Note: you must provide `BACKUP_FILENAME` without the file extension, e.g. `2020-04-22T00:00:01Z.87793a5e1556e0ff8a8b528328fab4ad`


Full example
------------

A complete docker-compose example for Wordpress (docker-compose-wp-example.yml) is available in the repository.

1. Open `docker-compose-wp-example.yml` and enter your S3 region, bucket and credentials
2. Run it with `docker-compose -f docker-compose-wp-example.yml up -d`
3. Open http://localhost:8080 in your browser and set up Wordpress.
4. See backup happening every 15 minutes in the logs: `docker-compose -f docker-compose-wp-example.yml logs --follow`
5. Force an immediate backup: `docker-compose -f docker-compose-wp-example.yml run --rm backup no-cron`
6. Mess with Wordpress...
7. Go to your AWS S3 console to pick up the latest backup filename
8. Restore your backup: `docker-compose -f docker-compose-wp-example.yml run --rm backup restore BACKUP_FILENAME`
9. Check that your backup is restored: http://localhost:8080


Security notes
--------------

To ensure your backup confidentiality, you should enable [Amazon S3 Block Public Access](https://aws.amazon.com/fr/s3/features/block-public-access/) feature.

If an attacker has compromised your container, your data is compromised but at least you should make sure he can't delete your backups. To do so, create an IAM strategy for the backup account to authorize __only put and get__ (but not list)
e.g.
```yaml
{
    "Version": "2012-10-17",
    "Statement": {
        "Effect": "Allow",
        "Action": [
            "s3:GetObject",
            "s3:PutObject"
        ],
        "Resource": [
            "arn:aws:s3:::your-bucket-1/*",
            "arn:aws:s3:::your-bucket-2/*",
            "arn:aws:s3:::your-bucket-3/*"
        ]
    }
}
```

Each backup file is appended a random string for uniqueness purposes. The random string makes the filename unguessable by an attacker, were the container compromised.  Thus, by limiting the S3 account permissions to put and get (but not list), the attacker couldn't overwrite old backup files.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [3.0.0] - 2023-03-04

- Added tags to target specific Postgres version (e.g. hleroy/backup-s3:v3.0-pg13)
- Added build script to automate building all images with relevant tags
- Added --if-exists to postgres dropdb statement to avoid throwing an error is the database doesn't exist. Thank you Matthew Vella for raising the issue.
- Reorganized Dockerfile to improve build time for subsequent build
- Fixed apt-key deprecation warning


## [2.2.0] - 2023-01-14

- Added option to use an alternative endpoint compatible with Amazon S3


## [2.1.0] - 2022-06-25

- Fix restore script assuming database name of "postgres". Thank you Nick Brassel for spotting the issue


## [2.0.0] - 2021-09-19

- **Breaking change** Added mandatory BACKUP_ENABLED environment var. If missing, container exits with code 0
- Added more checks for environment vars. If S3 or database settings are missing, container exits with code 1
- Added init: true in docker-compose.yml example to enable faster container termination


## [1.4.0] - 2021-08-31

- Use latest postgres client (v13)
- Postgres: dropdb before loading dump


## [1.3.0] - 2020-12-09

- Ignore tar warning "file changed as we read it"


## [1.2.0] - 2020-08-25

- Use latest postgres client (v12)


## [1.1.0] - 2020-08-18

- Fixed mysqldump error message which appeared with mysql 5.3.31:
  ```
  Error: 'Access denied; you need (at least one of) the PROCESS privilege(s) for this operation' when trying to dump tablespaces
  ```


## [1.0.0] - 2020-05-02

- Initial version


build.sh
#!/usr/bin/env bash
# vim: ai ts=4 sts=4 et sw=4
#
#     This file is part of hleroy/backup-s3. backup-s3 periodically backs up
#     a database (postgres or mysql) and/or data folder(s) to Amazon S3.
#     Copyright © 2012-2023 Hervé Le Roy
#
#     This program is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with this program.  If not, see <https://www.gnu.org/licenses/>.

VERSION="v3.0"

echo "Building hleroy/backup-s3 ${VERSION}"

docker build -t hleroy/backup-s3:${VERSION}-pg15 -t hleroy/backup-s3:latest  --build-arg POSTGRES_VERSION=15 .
docker build -t hleroy/backup-s3:${VERSION}-pg14 --build-arg POSTGRES_VERSION=14 .
docker build -t hleroy/backup-s3:${VERSION}-pg13 --build-arg POSTGRES_VERSION=13 .

docker images hleroy/backup-s3

start.sh
#!/usr/bin/env bash
# vim: ai ts=4 sts=4 et sw=4
#
#     This file is part of hleroy/backup-s3. backup-s3 periodically backs up
#     a database (postgres or mysql) and/or data folder(s) to Amazon S3.
#     Copyright © 2012-2023 Hervé Le Roy
#
#     This program is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with this program.  If not, see <https://www.gnu.org/licenses/>.


# Interrupt the script on error (set -e) and print an error message (trap)
set -eo pipefail
err() {
    echo "[CRITICAL] Backup container didn't start properly."
    awk 'NR>L-4 && NR<L+4 { printf "%-5d%3s%s\n",NR,(NR==L?">>>":""),$0 }' L=$1 $0
}
trap 'err $LINENO' ERR

# Catch the SIGINT and SIGTERM signal and then exit
trap "exit" SIGINT
trap "exit" SIGTERM

# Check if backup is enabled before starting
if [[ -z $BACKUP_ENABLED ]]; then
    echo "[INFO] Backup not enabled."
    echo "[INFO] Exiting in 5 seconds."
    sleep 5
    exit 0
else
    echo "[INFO] Backup enabled."
    echo "[INFO] Checking settings."
fi

# We need at least a path or a database to backup
if [[ -z $DATA_PATH && -z $DB_ENGINE ]]; then
    echo "[WARNING] Nothing to backup."
    exit 1
fi

# Check if S3 environment variable are set
if [[ -z $S3_ACCESS_KEY_ID || -z $S3_SECRET_ACCESS_KEY  || -z $S3_BUCKET || -z $S3_REGION ]]; then
    echo "[WARNING] Missing S3 environment variable(s)."
    : ${S3_ACCESS_KEY_ID?"You need to set the S3_ACCESS_KEY_ID environment variable."}
    : ${S3_SECRET_ACCESS_KEY?"You need to set the S3_SECRET_ACCESS_KEY environment variable."}
    : ${S3_BUCKET?"You need to set the S3_BUCKET environment variable."}
    : ${S3_REGION?"You need to set the S3_REGION environment variable."}
    exit 1
fi

# Check if database host, name and credentials are set
if [[ $DB_ENGINE == "postgres" || $DB_ENGINE == "mysql" ]]; then
    if [[ -z $DB_NAME || -z $DB_HOST  || -z $DB_USER || -z $DB_PASS ]]; then
        echo "[WARNING] Missing database environment variable(s)."
        : ${DB_NAME?"You need to set the DB_NAME environment variable."}
        : ${DB_HOST?"You need to set the DB_HOST environment variable."}
        : ${DB_USER?"You need to set the DB_USER environment variable."}
        : ${DB_PASS?"You need to set the DB_PASS environment variable."}
        exit 1
    fi
fi

# Read cron schedule from env var or provide default
CRON_SCHEDULE=${CRON_SCHEDULE:-0 0 * * *}

echo "[INFO] Settings OK."

if [[ "$1" == 'no-cron' ]]; then

    # Run backup now
    exec /backup.sh

elif [[ "$1" == 'restore' ]]; then

    # Restore backup
    exec /restore.sh $2

else

    # Schedule backup with cron

    # Create logfile
    LOGFIFO='/var/log/cron.fifo'
    if [[ ! -e "$LOGFIFO" ]]; then
        mkfifo "$LOGFIFO"
    fi

    # Save env vars
    printenv | sed 's/^\(.*\)\=\(.*\)$/export \1\="\2"/g' > env.sh

    # Setup crontab
    echo -e "SHELL=/bin/bash\nBASH_ENV=/env.sh\n$CRON_SCHEDULE /backup.sh > $LOGFIFO 2>&1" | crontab -

    # Print crontab
    echo "[INFO] The following crontab was configured:"
    crontab -l

    # Launch cron
    echo "[INFO] Launching cron"
    cron

    # Listen on the logs for changes
    exec tail -f "$LOGFIFO"
fi


restore.sh
#!/usr/bin/env bash
# vim: ai ts=4 sts=4 et sw=4
#
#     This file is part of hleroy/backup-s3. backup-s3 periodically backs up
#     a database (postgres or mysql) and/or data folder(s) to Amazon S3.
#     Copyright © 2012-2023 Hervé Le Roy
#
#     This program is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with this program.  If not, see <https://www.gnu.org/licenses/>.


# Interrupt the script on error (set -e) and print an error message (trap)
set -eo pipefail
err() {
    echo "[CRITICAL] Restore failed."
    awk 'NR>L-4 && NR<L+4 { printf "%-5d%3s%s\n",NR,(NR==L?">>>":""),$0 }' L=$1 $0
}
trap 'err $LINENO' ERR

# We need at least a path or a database to backup
if [[ -z $DATA_PATH && -z $DB_ENGINE ]]; then
    echo "[WARNING] Nothing to backup."
    exit 1
fi

# Check if S3 environment variable are set
if [[ -z $S3_ACCESS_KEY_ID || -z $S3_SECRET_ACCESS_KEY  || -z $S3_BUCKET || -z $S3_REGION ]]; then
    echo "[WARNING] Missing S3 environment variable(s)."
    : ${S3_ACCESS_KEY_ID?"You need to set the S3_ACCESS_KEY_ID environment variable."}
    : ${S3_SECRET_ACCESS_KEY?"You need to set the S3_SECRET_ACCESS_KEY environment variable."}
    : ${S3_BUCKET?"You need to set the S3_BUCKET environment variable."}
    : ${S3_REGION?"You need to set the S3_REGION environment variable."}
    exit 1
fi

# Check if database host, name and credentials are set
if [[ $DB_ENGINE == "postgres" || $DB_ENGINE == "mysql" ]]; then
    if [[ -z $DB_NAME || -z $DB_HOST  || -z $DB_USER || -z $DB_PASS ]]; then
        echo "[WARNING] Missing database environment variable(s)."
        : ${DB_NAME?"You need to set the DB_NAME environment variable."}
        : ${DB_HOST?"You need to set the DB_HOST environment variable."}
        : ${DB_USER?"You need to set the DB_USER environment variable."}
        : ${DB_PASS?"You need to set the DB_PASS environment variable."}
        exit 1
    fi
fi

# Check that a file name is supplied as first argument to this script
if [ -z "$1" ]; then
    echo "[ERROR] No file name supplied"
    exit 1
fi
BACKUPNAME=$1
echo "[INFO] Restoring $BACKUPNAME from $S3_REGION"

# Export environment vars needed for aws tools
export AWS_ACCESS_KEY_ID=$S3_ACCESS_KEY_ID
export AWS_SECRET_ACCESS_KEY=$S3_SECRET_ACCESS_KEY
export AWS_DEFAULT_REGION=$S3_REGION

# Restore database
if [[ $DB_ENGINE == "postgres" || $DB_ENGINE == "mysql" ]]; then

  # Fetch database backup
  echo "[INFO] Fetching database backup from S3"
  if [[ -z $S3_ENDPOINT_URL ]]; then
    aws s3 cp s3://$S3_BUCKET/$DB_ENGINE/$BACKUPNAME.sql.gz dump.sql.gz
  else
    aws s3 cp s3://$S3_BUCKET/$DB_ENGINE/$BACKUPNAME.sql.gz --endpoint-url=$S3_ENDPOINT_URL dump.sql.gz
  fi
  gzip -d dump.sql.gz

  echo "[INFO] Restoring database"
  if [ $DB_ENGINE == "postgres" ]; then

      # Restore postgres backup
      DB_PORT=${DB_PORT:-5432}
      export PGPASSWORD=$DB_PASS
      POSTGRES_HOST_OPTS="-h $DB_HOST -p $DB_PORT -U $DB_USER"
      #psql $POSTGRES_HOST_OPTS -d $DB_NAME -c "drop schema public cascade; create schema public;"
      dropdb --if-exists $POSTGRES_HOST_OPTS $DB_NAME
      pg_restore -C $POSTGRES_HOST_OPTS -d template1 < dump.sql

  elif [ $DB_ENGINE == "mysql" ]; then

      # Restore mysql backup
      DB_PORT=${DB_PORT:-3306}
      MYSQL_HOST_OPTS="--host=$DB_HOST --port=$DB_PORT --user=$DB_USER --password=$DB_PASS"
      mysql $MYSQL_HOST_OPTS $DB_NAME < dump.sql

  fi

fi


# Restore data
if [[ ! -z $DATA_PATH ]]; then

  # Fetch data backup
  echo "[INFO] Fetching data backup from S3"
  if [[ -z $S3_ENDPOINT_URL ]]; then
    aws s3 cp s3://$S3_BUCKET/data/$BACKUPNAME.tar data.tar
  else
    aws s3 cp s3://$S3_BUCKET/data/$BACKUPNAME.tar --endpoint-url=$S3_ENDPOINT_URL data.tar
  fi

  # Data path can contain multiple paths separated by colons
  IFS=: read -r -a data_path_array <<<"$DATA_PATH"

  # Remove content from directories (but not the directories themselves)
  for dir in "${data_path_array[@]}"; do
    echo "[INFO] Deleting content from $dir"
    rm -rf "$dir/*"
  done

  echo "[INFO] Restoring data"
  tar xf data.tar --absolute-names

fi

echo "[INFO] Restoration complete"


backup.sh
#!/usr/bin/env bash
# vim: ai ts=4 sts=4 et sw=4
#
#     This file is part of hleroy/backup-s3. backup-s3 periodically backs up
#     a database (postgres or mysql) and/or data folder(s) to Amazon S3.
#     Copyright © 2012-2023 Hervé Le Roy
#
#     This program is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with this program.  If not, see <https://www.gnu.org/licenses/>.


# Interrupt the script on error (set -e) and print an error message (trap)
set -eo pipefail
err() {
    echo "[CRITICAL] Backup failed."
    awk 'NR>L-4 && NR<L+4 { printf "%-5d%3s%s\n",NR,(NR==L?">>>":""),$0 }' L=$1 $0
}
trap 'err $LINENO' ERR

# We need at least a path or a database to backup
if [[ -z $DATA_PATH && -z $DB_ENGINE ]]; then
    echo "[WARNING] Nothing to backup."
    exit 1
fi

# Check if S3 environment variable are set
if [[ -z $S3_ACCESS_KEY_ID || -z $S3_SECRET_ACCESS_KEY  || -z $S3_BUCKET || -z $S3_REGION ]]; then
    echo "[WARNING] Missing S3 environment variable(s)."
    : ${S3_ACCESS_KEY_ID?"You need to set the S3_ACCESS_KEY_ID environment variable."}
    : ${S3_SECRET_ACCESS_KEY?"You need to set the S3_SECRET_ACCESS_KEY environment variable."}
    : ${S3_BUCKET?"You need to set the S3_BUCKET environment variable."}
    : ${S3_REGION?"You need to set the S3_REGION environment variable."}
    exit 1
fi

# Check if database host, name and credentials are set
if [[ $DB_ENGINE == "postgres" || $DB_ENGINE == "mysql" ]]; then
    if [[ -z $DB_NAME || -z $DB_HOST  || -z $DB_USER || -z $DB_PASS ]]; then
        echo "[WARNING] Missing database environment variable(s)."
        : ${DB_NAME?"You need to set the DB_NAME environment variable."}
        : ${DB_HOST?"You need to set the DB_HOST environment variable."}
        : ${DB_USER?"You need to set the DB_USER environment variable."}
        : ${DB_PASS?"You need to set the DB_PASS environment variable."}
        exit 1
    fi
fi

# Export environment vars needed for aws tools
export AWS_ACCESS_KEY_ID=$S3_ACCESS_KEY_ID
export AWS_SECRET_ACCESS_KEY=$S3_SECRET_ACCESS_KEY
export AWS_DEFAULT_REGION=$S3_REGION

# Format backup date
BACKUP_DATE=`date +"%Y-%m-%dT%H:%M:%SZ"`

# Generate random string for uniqueness purposes and so that you can limit the AWS user
# to only be able to put and get (not list) and thus the random string makes it so that,
# were the machine compromised, that AWS account couldn't overwrite old backup files
RAND_STR=$(cat /proc/sys/kernel/random/uuid)

# Dump database
if [[ $DB_ENGINE == "postgres" || $DB_ENGINE == "mysql" ]]; then

  echo "[INFO] Creating dump of ${DB_NAME} database"

  if [ $DB_ENGINE == "postgres" ]; then

      # Create postgres backup
      DB_PORT=${DB_PORT:-5432}
      export PGPASSWORD=$DB_PASS
      POSTGRES_HOST_OPTS="-h $DB_HOST -p $DB_PORT -U $DB_USER"
      pg_dump -Fc $POSTGRES_HOST_OPTS $DB_NAME | gzip > dump.sql.gz

  elif [ $DB_ENGINE == "mysql" ]; then

      # Create mysql backup
      DB_PORT=${DB_PORT:-3306}
      MYSQL_HOST_OPTS="--host=$DB_HOST --port=$DB_PORT --user=$DB_USER --password=$DB_PASS"
      mysqldump --opt --add-drop-database --no-tablespaces $MYSQL_HOST_OPTS $DB_NAME | gzip > dump.sql.gz

  fi

  # Upload database backup to Amazon S3 or alternative endpoint
  echo "[INFO] Uploading database backup to $S3_BUCKET"
  if [[ -z $S3_ENDPOINT_URL ]]; then
    cat dump.sql.gz | aws s3 cp - s3://$S3_BUCKET/$DB_ENGINE/$BACKUP_DATE.$RAND_STR.sql.gz || exit 2
  else
    cat dump.sql.gz | aws s3 cp - s3://$S3_BUCKET/$DB_ENGINE/$BACKUP_DATE.$RAND_STR.sql.gz --endpoint-url=$S3_ENDPOINT_URL || exit 2
  fi

fi

# Create data backup
if [[ ! -z $DATA_PATH ]]; then

  # Data path can contain multiple paths separated by colons
  IFS=: read -r -a data_path_array <<<"$DATA_PATH"

  FILE='data.tar'
  # Remove file if it exists
  if [ -f $FILE ] ; then
      rm $FILE
  fi

  echo "[INFO] Creating tar archive"
  for dir in "${data_path_array[@]}"
  do
     echo "[INFO] Adding $dir"
     # tar sometime fails with the error message "file changed as we read it" and it exits with an error code 1
     # the purpose of the code after the tar command is to ignore code 1 and return all other codes
     # Credit: https://stackoverflow.com/questions/20318852/tar-file-changed-as-we-read-it
     tar rf $FILE --warning=no-file-changed --absolute-names "$dir" || ( export ret=$?; [[ $ret -eq 1 ]] || exit "$ret" )
  done

  # Upload data backup to Amazon S3 or alternative endpoint
  echo "[INFO] Uploading data backup to $S3_BUCKET"
  if [[ -z $S3_ENDPOINT_URL ]]; then
    cat $FILE | aws s3 cp - s3://$S3_BUCKET/data/$BACKUP_DATE.$RAND_STR.tar || exit 2
  else
    cat $FILE | aws s3 cp - s3://$S3_BUCKET/data/$BACKUP_DATE.$RAND_STR.tar --endpoint-url=$S3_ENDPOINT_URL || exit 2
  fi

fi

echo "[INFO] Backup uploaded successfully to S3"
